gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java

```diff
@@ -0,0 +1,296 @@
+        break;
+      case NULL:
+        flattenedSchema = Schema.create(schema.getType());
+        break;
+      case RECORD:
+        List<Schema.Field> flattenedFields = new ArrayList<>();
+        if (null != schema.getFields() && schema.getFields().size() > 0) {
+          // Recursively flatten fields
+          for (Schema.Field oldField : schema.getFields()) {
+            LinkedList<String> lineage = new LinkedList<>();
+            lineage.add(schema.getName());
+            List<Schema.Field> newFields = flattenField(oldField, lineage, level+1, false);
+            if (null != newFields && newFields.size() > 0) {
+              flattenedFields.addAll(newFields);
+            }
+          }
+        }
+        flattenedSchema = Schema.createRecord(schema.getName(), schema.getDoc(), schema.getNamespace(),
+            schema.isError());
+        flattenedSchema.setFields(flattenedFields);
+        break;
+      case STRING:
+        flattenedSchema = Schema.create(schema.getType());
+        break;
+      case UNION:
+        List<Schema> flattenedUnionMembers = new ArrayList<>();
+        if (null != schema.getTypes() && schema.getTypes().size() > 0) {
+          for (Schema oldUnionMember : schema.getTypes()) {
+            // Its member might still recursively contain records
+            flattenedUnionMembers.add(flatten(oldUnionMember, level + 1));
+          }
+        }
+        flattenedSchema = Schema.createUnion(flattenedUnionMembers);
+        break;
+      default:
+        String exceptionMessage = String.format("Schema flattening failed for \"%s\" ", schema);
+        LOG.error(exceptionMessage);
+
+        throw new AvroRuntimeException(exceptionMessage);
+    }
+
+    // Copy schema metadata
+    copyProperties(schema, flattenedSchema);
+
+    return flattenedSchema;
+  }
+
+  /***
+   * Flatten Record field, and compute a list of flattened fields
+   * @param f Field to flatten
+   * @param lineage Lineage represented as a List of Strings
+   * @param level Recursion level, if it varies that size of lineage - it indicates a flattening and the fields are
+   *              decorated with lineage information
+   * @param shouldWrapInOption If the field should be wrapped as an OPTION, if we un-nest fields within an OPTION,
+   *                           we make all the unnested fields as OPTIONs
+   * @return List of flattened Record fields
+   */
+  private static List<Schema.Field> flattenField(Schema.Field f, LinkedList<String> lineage, int level,
+      boolean shouldWrapInOption) {
+    Preconditions.checkNotNull(f);
+    Preconditions.checkNotNull(f.schema());
+    Preconditions.checkNotNull(f.name());
+
+    List<Schema.Field> flattenedFields = new ArrayList<>();
+
+    // Set lineage
+    if (null == lineage) {
+      lineage = new LinkedList<>();
+    }
+    lineage.add(f.name());
+
+    // If field.Type = RECORD, un-nest its fields and return them
+    if (Schema.Type.RECORD.equals(f.schema().getType())) {
```

Comment:
This seems to be similar logic as the `case RECORD` in line 95. Can you put it in a method `flattenRecord`?